jlm/hls/backend/rhls2firrtl/RhlsToFirrtlConverter.cpp

```diff
@@ -784,84 +790,114 @@ RhlsToFirrtlConverter::MlirGenHlsMemReq(const jlm::rvsdg::simple_node * node)
 
   auto zeroBitValue = GetConstant(body, 1, 0);
   auto oneBitValue = GetConstant(body, 1, 1);
-  // default req connection
-  Connect(body, memReqValid, zeroBitValue);
-  ConnectInvalid(body, memReqBundle);
-  mlir::Value previousGranted = GetConstant(body, 1, 0);
-  for (size_t i = 0; i < loadTypes->size(); ++i)
-  {
-    Connect(body, loadAddrReadys[i], zeroBitValue);
-    auto notOp = AddNotOp(body, previousGranted);
-    auto condition = AddAndOp(body, notOp, loadAddrValids[i]);
-    auto whenOp = AddWhenOp(body, condition, false);
-    auto thenBody = whenOp.getThenBodyBuilder().getBlock();
-    Connect(thenBody, loadAddrReadys[i], memReqReady);
-    Connect(thenBody, memReqValid, loadAddrValids[i]);
-    Connect(thenBody, memReqAddr, loadAddrDatas[i]);
-    Connect(thenBody, memReqId, loadIds[i]);
-    // no data or write
-    int bitWidth;
-    auto loadType = loadTypes->at(i).get();
-    if (auto bitType = dynamic_cast<const jlm::rvsdg::bittype *>(loadType))
-    {
-      bitWidth = bitType->nbits();
-    }
-    else if (dynamic_cast<const jlm::llvm::PointerType *>(loadType))
-    {
-      bitWidth = 64;
-    }
-    else
-    {
-      throw jlm::util::error("unknown width for mem request");
-    }
-    int log2Bytes = log2(bitWidth / 8);
-    Connect(thenBody, memReqSize, GetConstant(thenBody, 3, log2Bytes));
+  ::llvm::SmallVector<mlir::Value> loadGranted(loadTypes->size(), zeroBitValue);
+  ::llvm::SmallVector<mlir::Value> storeGranted(storeTypes->size(), zeroBitValue);
+  for (size_t j = 0; j < node->noutputs(); ++j)
+  {
+    auto reqType = dynamic_cast<const jlm::hls::bundletype *>(&node->output(j)->type());
+    // TODO: more robust check
+    auto hasWrite = reqType->elements_.size() == 5;
```

Comment:
Can it never happen that the type of the output is not a bundletype?

Reply:
If that is the case (as the code suggest), then maybe perform a `AssertedCast<>` instead. Otherwise, a check is needed.

Reply:
Changed to AssertedCast